pdns/dnsdistdist/dnsdist-async.cc

```diff
@@ -79,10 +81,13 @@ void AsynchronousHolder::stop()
   notify();
 }
 
+// NOLINTNEXTLINE(performance-unnecessary-value-param): this is a long-lived thread, and we want to make sure the reference count of the shared pointer has been increased
```

Comment:
I find it strange that clang-tidy complains about passing a shared pointer object by value...

Reply:
That's because to do that you need to increment the reference counter, which is not cheap, while taking a reference to the same `shared_pointer` is free.

Reply:
Sure. But that's the whole point of using a `shared_ptr`. When I pass it by value I really mean it, so I find it stupid that `clang-tidy` is complaining about this and suggesting a reference instead.

Reply:
I'm a bit torn, to be honest, because in my experience passing a `shared_pointer` by value is actually very often an oversight. On the other hand it's annoying to have to justify yourself every time you do so on purpose.

Reply:
Is the ref counter incremented by passing a `shared_ptr` as a reference?

Reply:
No, it's not, you get the same `shared_ptr`, and not a different `shared_ptr` pointing to the same data and metadata.

Reply:
Well in that case I would disagree that passing it by value would be an oversight. If a reference needs to be passed then it needs to be a reference to the internal object rather than the shared pointer itself, this way it's clear that the caller has to outlive the callee.

Reply:
For maximum safety, yes, for maximum performance no :)

Reply:
I think getting a reference to the internal object from the shared pointer object is very cheap... no?

Reply:
You are right, of course! I read your previous comment too quickly, sorry! I agree it's better to pass `const T&` than `const std::shared_ptr<const T>&`, clearly!